File: bazelrio/scripts/deploy/deploy.py
import argparse
import hashlib
import os
import sys
from contextlib import contextmanager
from os.path import basename
from shlex import quote

from alive_progress import alive_bar
from blessed import Terminal
from paramiko.client import MissingHostKeyPolicy, SSHClient
from rules_python.python.runfiles import runfiles

DEPLOYED_FILE_PERMS = 0o755
DYLIB_DIR = "/usr/local/frc/third-party/lib/"
LVUSER_UID = 500
LVUSER_GID = 500
ROBOT_COMMAND_PATH = "/home/lvuser/robotCommand"

term = Terminal()

class SilentAllowPolicy(MissingHostKeyPolicy):
    def missing_host_key(self, client, hostname, key):
        return


def establish_connection(team_number, verbose):
    # addresses supplied from
    # https://github.com/wpilibsuite/GradleRIO/blob/aaf4da44e914a49e0bb956b028130481f538b31c/src/main/groovy/edu/wpi/first/gradlerio/frc/RoboRIO.groovy#L40
    addresses = [
        f"roborio-{team_number}-frc.local",
        f"10.{team_number // 100}.{team_number % 100}.2",
       "172.22.11.2",
        f"roborio-{team_number}-frc",
        f"roborio-{team_number}-frc.lan",
        f"roborio-{team_number}-frc.frc-field.local",
    ]

    client = SSHClient()
    client.set_missing_host_key_policy(SilentAllowPolicy)

    for address in addresses:
        print(term.bright_yellow(f"Attempting to connect to {address}..."))
        try:
            client.connect(
                address,
                username="admin",
                password="",
                look_for_keys=False,
                allow_agent=False,
                timeout=15,
            )
            print(term.bright_green(f"Connected to {address}."))
            return client
        except Exception as e:
            if verbose:
                print(e)
            print(term.bright_red("Connection failed."))
    print(term.bright_red("All attempts to connect to the roboRIO failed. Exiting."))
    sys.exit(1)


def transfer_file(ssh_client, sftp_client, local_fo, remote_path, verbose):
    local_size = os.fstat(local_fo.fileno()).st_size

    if verbose:
        print(f"checking {remote_path}")

    try:
        remote_fo = sftp_client.open(remote_path)
        remote_size = remote_fo.stat().st_size

        if verbose:
            print(f"local size: {local_size}, remote size: {remote_size}")

        if remote_size == local_size:
            local_hash = hashlib.md5(local_fo.read()).hexdigest()
            _, remote_hash_stdout_fo, _ = ssh_client.exec_command(f"md5sum {quote(remote_path)}")
            remote_hash = remote_hash_stdout_fo.read().split()[0].decode()

            if verbose:
                print(f"local hash: {local_hash}, remote hash: {remote_hash}")

            if remote_hash == local_hash:
                return
    except Exception as e:
        # could fail due to misisng file, misisng md5sum, etc.
        if verbose:
            print(e)
        pass

    if verbose:
        print(f"copying {local_fo.name} -> {remote_path}")

    sftp_client.putfo(local_fo, remote_path)


class WindowsProgressBar:
    def __init__(self, total):
        self.total = total
        self.current = 1

    def __call__(self):
        self.current += 1
        pass

    def text(self, msg):
        print(f"({self.current}/{self.total}) {msg}")


@contextmanager
def get_progress_bar(total):
    if os.name != "nt":
        with alive_bar(
            total,
            enrich_print=False,
            monitor=False,
            stats=False,
            theme="classic",
        ) as progress_bar:
            yield progress_bar
    else:
        yield WindowsProgressBar(total)


def deploy(argv):
    parser = argparse.ArgumentParser(description="Deploy code to a roboRIO")
    parser.add_argument("--robot_binary", type=str, required=True)
    parser.add_argument("--robot_command", type=str, default="{}")
    parser.add_argument("--team_number", type=int, required=True)
    parser.add_argument("--verbose", action="store_true", default=False)
    parser.add_argument("--dynamic_libraries", nargs='*', default=[])
    args = parser.parse_args(argv)

    with get_progress_bar(len(args.dynamic_libraries) + 1) as progress_bar:
        r = runfiles.Create()

        client = establish_connection(args.team_number, args.verbose)
        sftp_client = client.open_sftp()

        binary_name = basename(args.robot_binary)
        destination_path = f"/home/lvuser/{binary_name}"
        quoted_destination_path = quote(destination_path)

        print(term.bright_yellow(f"Attempting to deploy {binary_name}..."))

        # deploy process adapted from
        # https://github.com/wpilibsuite/GradleRIO/blob/aaf4da44e914a49e0bb956b028130481f538b31c/src/main/groovy/edu/wpi/first/gradlerio/frc/FRCNativeArtifact.groovy#L29

        # stop and remove existing robot binary
        client.exec_command(". /etc/profile.d/natinst-path.sh; /usr/local/frc/bin/frcKillRobot.sh -t")
        try:
            sftp_client.remove(destination_path)
        except FileNotFoundError as e:
            if args.verbose:
                print(e)

        # copy new robot binary
        progress_bar.text(binary_name)
        with open(r.Rlocation(os.path.join(__name__, args.robot_binary)), "rb") as robot_binary_fo:
            transfer_file(client, sftp_client, robot_binary_fo, destination_path, args.verbose)
        sftp_client.chmod(destination_path, DEPLOYED_FILE_PERMS)
        sftp_client.chown(destination_path, LVUSER_UID, LVUSER_GID)
        client.exec_command(f"setcap cap_sys_nice+eip {quoted_destination_path}")
        progress_bar()

        # write new robotCommand
        with sftp_client.open(ROBOT_COMMAND_PATH, "w") as fo:
            # we take a robotCommand format string as a argument to make it easier for different languages (ie java) to
            # describe how their binaries should be executed on the rio.
            # the remote location of the robot binary is substituted into the format string, and the whole thing is
            # wrapped in a bash command that sets -x before execing into the formatted string.
            # this is done as a convenience so that the user can see exactly what is running in FRC_UserProgram.log.
            inner_robot_command = args.robot_command.format(quoted_destination_path)
            bash_command = f"set -euxo pipefail; exec {inner_robot_command}"
            fo.write(f"bash -c {quote(bash_command)}\n")
        sftp_client.chmod(ROBOT_COMMAND_PATH, DEPLOYED_FILE_PERMS)
        sftp_client.chown(ROBOT_COMMAND_PATH, LVUSER_UID, LVUSER_GID)

        # copy shared libraries
        try:
            sftp_client.mkdir(DYLIB_DIR)
        except IOError as e:
            if args.verbose:
                print(e)
        for dylib_path in args.dynamic_libraries:
            dylib_name = basename(dylib_path)
            progress_bar.text(dylib_name)
            with open(r.Rlocation(os.path.join(__name__, dylib_path)), "rb") as dylib_fo:
                transfer_file(client, sftp_client, dylib_fo, f"{DYLIB_DIR}{dylib_name}", args.verbose)
            progress_bar()

        # restart robot code
        client.exec_command("sync")
        client.exec_command("ldconfig")
        client.exec_command(". /etc/profile.d/natinst-path.sh; /usr/local/frc/bin/frcKillRobot.sh -t -r")

        print(term.bright_green(f"Deployed {binary_name}. Exiting."))


if __name__ == "__main__":
    deploy(sys.argv[1:])

File: bazelrio/scripts/deploy/Deploy.java
package org.bazelrio.deploy;

import com.google.devtools.build.runfiles.Runfiles;
import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.List;
import me.tongfei.progressbar.ProgressBar;
import me.tongfei.progressbar.ProgressBarBuilder;
import me.tongfei.progressbar.ProgressBarStyle;
import net.schmizz.sshj.SSHClient;
import net.schmizz.sshj.connection.channel.direct.Session;
import net.schmizz.sshj.connection.channel.direct.Session.Command;
import net.schmizz.sshj.transport.verification.HostKeyVerifier;
import net.schmizz.sshj.xfer.FileSystemFile;
import net.schmizz.sshj.xfer.scp.SCPUploadClient;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.impl.Arguments;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;

class NoopKeyVerifier implements HostKeyVerifier {
  public List<String> findExistingAlgorithms(String _hostname, int _port) {
    return new ArrayList<>();
  }

  public boolean verify(String _hostname, int _port, PublicKey _key) {
    return true;
  }
}

class CommandFailedException extends IOException {
  public CommandFailedException(String message) { super(message); }
}

class Deploy {
  public static void main(String[] args) throws IOException {
    System.out.println("Starting deploy");

    ArgumentParser parser =
        ArgumentParsers.newFor("deploy").build().description(
            "Deploy code to a roboRIO");
    parser.addArgument("--robot_binary")
        .type(String.class)
        .help("the executable binary to deploy")
        .required(true);
    parser.addArgument("--robot_command")
        .type(String.class)
        .help(
            "the script to run the robot binary (deployed to /home/lvuser/robotCommand)")
        .setDefault("{}");
    parser.addArgument("--team_number")
        .type(Integer.class)
        .help("the team number, used for the roboRIO addresses")
        .required(true);
    parser.addArgument("--verbose")
        .action(Arguments.storeTrue())
        .setDefault(false);
    parser.addArgument("--dynamic_libraries")
        .type(String.class)
        .nargs("*")
        .help("the .so libraries which the executable requires")
        .setDefault();
    parser.addArgument("--skip_dynamic_libraries")
        .action(Arguments.storeTrue())
        .help(
            "don't deploy the dynamic libraries (warning: unsafe, might lead to unexpected issues)")
        .setDefault(false);

    Namespace parsedArgs = parser.parseArgsOrFail(args);

    Runfiles runfiles = Runfiles.create();

    File robotBinary = runfile(runfiles, parsedArgs.get("robot_binary"));
    String robotCommand = parsedArgs.get("robot_command");
    ArrayList<String> dynamicLibraryPaths = parsedArgs.get("dynamic_libraries");
    boolean verbose = parsedArgs.get("verbose");

    if (parsedArgs.get("skip_dynamic_libraries")) {
      if (verbose) {
        System.out.println("Skipping dynamic libraries");
      }
      dynamicLibraryPaths = new ArrayList<>();
    }

    String robotBinaryDestination =
        String.format("/home/lvuser/%s", robotBinary.getName());

    ProgressBar progressBar = new ProgressBarBuilder()
                                  .setTaskName("Deploying")
                                  .setInitialMax(dynamicLibraryPaths.size() + 1)
                                  .setStyle(ProgressBarStyle.ASCII)
                                  .setUpdateIntervalMillis(100)
                                  .build();
    progressBar.stepTo(0);

    SSHClient client = new SSHClient();
    client.addHostKeyVerifier(new NoopKeyVerifier());
    client.useCompression();

    if (!establishSession(client, parsedArgs.get("team_number"), verbose)) {
      System.err.println("Couldn't find a roboRIO");
      System.exit(-1);
    }

    SCPUploadClient scp = client.newSCPFileTransfer().newSCPUploadClient();

    // Stop and remove existing robot binary
    runCommand(
        client,
        ". /etc/profile.d/natinst-path.sh; /usr/local/frc/bin/frcKillRobot.sh -t",
        verbose);
    runCommand(client, String.format("rm -f %s", robotBinaryDestination),
               verbose);

    // Fix some files (required only once after roboRIO format)
    runCommand(
        client,
        "sed -i -e 's/\\\"exec /\\\"/' /usr/local/frc/bin/frcRunRobot.sh",
        verbose);
    runCommand(
        client,
        "sed -i -e 's/^StartupDLLs/;StartupDLLs/' /etc/natinst/share/ni-rt.ini",
        verbose);

    // Copy new robot binary
    progressBar.setExtraMessage(robotBinary.getName());
    scp.copy(new FileSystemFile(robotBinary), robotBinaryDestination);
    runCommand(client, String.format("chmod +x %s", robotBinaryDestination),
               verbose);
    runCommand(client,
               String.format("chown lvuser:ni %s", robotBinaryDestination),
               verbose);
    runCommand(
        client,
        String.format("setcap cap_sys_nice+eip %s", robotBinaryDestination),
        verbose);
    progressBar.step();

    // Write new robotCommand
    runCommand(
        client,
        String.format("echo %s > /home/lvuser/robotCommand",
                      robotCommand.replace("{}", robotBinaryDestination)),
        verbose);
    runCommand(client, "chmod +x /home/lvuser/robotCommand", verbose);
    runCommand(client, "chown lvuser:ni /home/lvuser/robotCommand", verbose);

    // Copy dynamic libraries
    for (String dynamicLibraryPath : dynamicLibraryPaths) {
      File dynamicLibrary = runfile(runfiles, dynamicLibraryPath);
      progressBar.setExtraMessage("Deploying " + dynamicLibrary.getName());
      String dynamicLibraryDestination = String.format(
          "/usr/local/frc/third-party/lib/%s", dynamicLibrary.getName());
      scp.copy(new FileSystemFile(dynamicLibrary), dynamicLibraryDestination);
      progressBar.step();
    }
    progressBar.close();

    // Restart robot code
    System.out.print("Restarting robot code... ");
    runCommand(client, "sync", verbose);
    runCommand(client, "ldconfig", verbose);
    runCommand(
        client,
        ". /etc/profile.d/natinst-path.sh; /usr/local/frc/bin/frcKillRobot.sh -t -r",
        verbose);
    System.out.println("Done.");
    System.out.print("Deploy completed!");
  }

  static File runfile(Runfiles runfiles, String location) {
    Path runfilePath = Path.of("__main__").resolve(location).normalize();
    ArrayList<String> pathParts = new ArrayList<>();
    runfilePath.iterator().forEachRemaining(
        part -> pathParts.add(part.toString()));
    return new File(runfiles.rlocation(String.join("/", pathParts)));
  }

  static void runCommand(SSHClient client, String commandString,
                         boolean verbose) throws IOException {
    if (verbose) {
      System.out.println(String.format("Running %s", commandString));
    }
    Command command = client.startSession().exec(commandString);
    command.join();
    int exitStatus = command.getExitStatus();
    if (exitStatus != 0) {
      throw new CommandFailedException(String.format(
          "Command %s exited with code %d", commandString, exitStatus));
    }
  }

  static boolean establishSession(SSHClient client, int teamNumber,
                                  boolean verbose) {
    String[] addresses = {
        String.format("roborio-%d-frc.local", teamNumber),
        String.format("10.%d.%d.2", teamNumber / 100, teamNumber % 100),
        "172.22.11.2",
        String.format("roborio-%d-frc", teamNumber),
        String.format("roborio-%d-frc.lan", teamNumber),
        String.format("roborio-%d-frc.frc-field.local", teamNumber),
    };

    ProgressBar progressBar = new ProgressBarBuilder()
                                  .setTaskName("roboRIO Search")
                                  .setInitialMax(addresses.length)
                                  .setStyle(ProgressBarStyle.ASCII)
                                  .setUpdateIntervalMillis(100)
                                  .build();
    progressBar.stepTo(0);

    for (String address : addresses) {
      try {
        progressBar.setExtraMessage(address);
        progressBar.step();
        if (verbose) {
          System.out.println(
              String.format("Attempting to connect to %s", address));
        }
        client.connect(address);
        client.authPassword("admin", "");

        progressBar.stepTo(addresses.length);
        progressBar.close();
        return true;
      } catch (IOException e) {
        if (verbose) {
          System.err.println(
              String.format("Error connecting to %s: %s", address, e));
        }
      }
    }

    progressBar.close();
    return false;
  }
}
